aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -0,0 +1,414 @@
+    hex_repr_len + CRLF.len() as u64 + payload_length + CRLF.len() as u64
+}
+
+fn trailers_as_aws_chunked_bytes(
+    total_length_of_trailers_in_bytes: u64,
+    trailer_map: Option<HeaderMap>,
+) -> Bytes {
+    use std::fmt::Write;
+
+    // On 32-bit operating systems, we might not be able to convert the u64 to a usize, so we just
+    // use `String::new` in that case.
+    let mut trailers = match usize::try_from(total_length_of_trailers_in_bytes) {
+        Ok(total_length_of_trailers_in_bytes) => {
+            String::with_capacity(total_length_of_trailers_in_bytes)
+        }
+        Err(_) => String::new(),
+    };
+    let mut already_wrote_first_trailer = false;
+
+    if let Some(trailer_map) = trailer_map {
+        for (header_name, header_value) in trailer_map.into_iter() {
+            match header_name {
+                // New name, new value
+                Some(header_name) => {
+                    if already_wrote_first_trailer {
+                        // First trailer shouldn't have a preceding CRLF, but every trailer after it should
+                        trailers.write_str(CRLF).unwrap();
+                    } else {
+                        already_wrote_first_trailer = true;
+                    }
+
+                    trailers.write_str(header_name.as_str()).unwrap();
+                    trailers.write_char(':').unwrap();
+                }
+                // Same name, new value
+                None => {
+                    trailers.write_char(',').unwrap();
+                }
+            }
+            trailers.write_str(header_value.to_str().unwrap()).unwrap();
+        }
+    }
+
+    // Write CRLF to end the body
+    trailers.write_str(CRLF).unwrap();
+    // If we wrote at least one trailer, we need to write an extra CRLF
+    if total_length_of_trailers_in_bytes != 0 {
+        trailers.write_str(CRLF).unwrap();
+    }
+
+    trailers.into()
+}
+
+impl Body for AwsChunkedBody<Inner> {
+    type Data = Bytes;
+    type Error = aws_smithy_http::body::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        tracing::info!("polling AwsChunkedBody");
+        let mut this = self.project();
+
+        match *this.state {
+            AwsChunkedBodyState::WritingChunkSize => match this.inner.poll_data(cx) {
+                Poll::Ready(Some(Ok(data))) => {
+                    // A chunk must be prefixed by chunk size in hexadecimal
+                    tracing::info!("writing chunk size and start of chunk");
+                    *this.state = AwsChunkedBodyState::WritingChunk;
+                    let total_chunk_size = this
+                        .options
+                        .chunk_length
+                        .or(this.options.stream_length)
+                        .unwrap_or_default();
+                    Poll::Ready(Some(Ok(prefix_with_chunk_size(data, total_chunk_size))))
```

Comment:
this adds a significant amount of copying—how hard would it be to refactor it to send the chunk size and the chunk separately?

I'm also a little confused about how this works—we never switch back into `WritingChunkSize` AFAICT